Move `dim` from schemes to rings

paemurru opened the pull request:
Contributes to #2771 but does not close it because https://github.com/oscar-system/Oscar.jl/issues/2771#issuecomment-1737280051

paemurru committed:
Move `dim` from schemes to rings

paemurru committed:
Add missing `return`

afkafkafk13 commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2565,7 +2565,41 @@ function small_generating_set(
+  error("Not implemented")
+end
+
+@attr Int function dim(R::MPolyQuoLocRing{<:Any, <:Any, <:MPolyRing, <:MPolyRingElem, <:Union{MPolyComplementOfPrimeIdeal, MPolyComplementOfKPointIdeal}})
+  P = prime_ideal(inverted_set(R))
+  I = saturated_ideal(modulus(R))
+  return dim(I) - dim(P)
+end
```
Please handle them separately. For MPolyComplementOfKPointIdeal P is certified to be zero-dimensional so that we do not need to compute dim(P).

paemurru replied:
> P is certified to be zero-dimensional

It seems that this is not true.

```
julia> R, (x, y) = ZZ[:x, :y]
(Multivariate polynomial ring in 2 variables over ZZ, ZZMPolyRingElem[x, y])

julia> S = complement_of_point_ideal(R, [0, 0])
Complement
  of maximal ideal corresponding to rational point with coordinates (0, 0)
  in multivariate polynomial ring in 2 variables over ZZ

julia> I = ideal(R, 0)
Ideal generated by
  0

julia> L, _ = localization(R, S)
(Localization of multivariate polynomial ring in 2 variables over ZZ at complement of maximal ideal of point (0, 0), Hom: R -> localized ring)

julia> W, _ = quo(L, L(I))
(Localization of quotient of multivariate polynomial ring at complement of maximal ideal, hom: Localized ring -> Localized quotient of multivariate polynomial ring)

julia> P = prime_ideal(inverted_set(W))
Ideal generated by
  x
  y

julia> dim(P)
1
```

So I will leave it as it is for now.

simonbrandhorst replied:
My guess is that `complement_of_point_ideal` silently assumes that we are working over a field? Perhaps the type and constructor should be restricted accordingly what do you think @afkafkafk13 ?

afkafkafk13 replied:
> julia> S = complement_of_point_ideal(R, [0, 0])
> Complement
>   of maximal ideal corresponding to rational point with coordinates (0, 0)
>   in multivariate polynomial ring in 2 variables over ZZ

This is indeed a problem with complement_of_point_ideal: 
This ideal is NOT maximal as we do not get a field as the quotient of the ring by it. This needs to be handled in `complement_of_point_ideal`. My suggestion: check dimension (not number of entries!!) of the ideal before doing anything else in `complement_of_point_ideal`, throw an error if it is not maximal (i.e. dim(ideal) is not zero -- this should already be working in Oscar) and point the user to `complement_of_prime_ideal`.

simonbrandhorst replied:
It is a bit fuzzy what a "point" is here. Will we get anything computationally effective for the "point" given by the maximal ideal `(2,x,y)` in `Z[x,y]`? I impression was that `ComplementOfKPointIdeal` just assumes a base field.

paemurru replied:
> I impression was that ComplementOfKPointIdeal just assumes a base field.

I think it does assume but does not check for this (although it should).

I would leave fixing `complement_of_point_ideal` to a separate issue.

afkafkafk13 replied:
ok. fine with me. But we urgently need to fix complement_of_point_ideal. Maybe we can just check

@simonbrandhorst no, we won't get anything computationally effective currently. You are right. We should just tell the user to use `complement_of_prime_ideal` instead, whenever we are not over a field.

Edit: This is a strategic decision and breaks some tests. I opened PR We need to discuss this. Over a ring this is not a point ideal and hence should not be used via this localization, anyway. Moreover, the localization at <x_1,\dots,x_n> in ZZ[x_1,\dots,x_n] is not what is produced here, as we are localizing w.r.t. a monomial ordering and continue working over ZZ and do not pass to QQ, even though ZZ \cap <x_1,\dots,x_n> = 0.

afkafkafk13 commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2565,7 +2565,41 @@ function small_generating_set(
+@attr Int function dim(R::MPolyLocRing{<:Any,<:Any,<:MPolyRing,<:MPolyRingElem, <:MPolyPowersOfElement})
+  # zariski open subset of A^n
+  return dim(closure(spec(R)))
+end
```
You are again passing through the scheme side here. Please try to use saturated ideal instead of closure. It should also have a variant for MPolyPowersOfElement.

afkafkafk13 commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2565,7 +2565,41 @@ function small_generating_set(
+@attr Int function dim(R::MPolyLocRing{<:Any,<:Any,<:MPolyRing,<:MPolyRingElem, <:Union{MPolyComplementOfPrimeIdeal, MPolyComplementOfKPointIdeal}})
+  P = prime_ideal(inverted_set(R))
+  return codim(P)
+end
```
Again you might want to handle MPolyComplementOfKPointIdeal separately?

paemurru replied:
I am not sure how to do this here.

simonbrandhorst commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2565,7 +2565,41 @@ function small_generating_set(
+
+@attr Int function dim(R::MPolyLocRing{<:Any,<:Any,<:MPolyRing,<:MPolyRingElem, <:MPolyPowersOfElement})
+  # zariski open subset of A^n
+  return dim(closure(spec(R)))
```
```suggestion
  return dim(base_ring(R))
```

simonbrandhorst commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2565,7 +2565,41 @@ function small_generating_set(
+end
+
+@attr Int function dim(R::MPolyQuoLocRing{<:Any, <:Any, <:MPolyRing, <:MPolyRingElem, <:MPolyPowersOfElement})
+  return dim(closure(spec(R)))
```
```suggestion
  return dim(saturated_ideal(R))
```

simonbrandhorst reviewed (commented):
Thanks for looking into this.

paemurru committed:
Separate dim for MPolyComplementOfKPointIdeal

Anne: "For MPolyComplementOfKPointIdeal P is certified to be zero-dimensional so that we do not need to compute dim(P)."

paemurru committed:
Apply suggestions from code review

Co-authored-by: Simon Brandhorst <brandhorst@math.uni-sb.de>

paemurru committed:
Remove redundant functions

paemurru committed:
Fix bug when base ring is the integers

(The bug was introduced and fixed in this pull request.)

paemurru committed:
dim for zzModRing and ZZModRing

simonbrandhorst commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -2565,7 +2565,29 @@ function small_generating_set(
+@attr Int function dim(R::MPolyLocRing{<:Any,<:Any,<:MPolyRing,<:MPolyRingElem, <:Union{MPolyComplementOfPrimeIdeal, MPolyComplementOfKPointIdeal}})
+  P = prime_ideal(inverted_set(R))
+  return codim(P)
+end
```
```suggestion
@attr Int function dim(R::MPolyLocRing{<:Any,<:Any,<:MPolyRing,<:MPolyRingElem, <:MPolyComplementOfPrimeIdeal})
  P = prime_ideal(inverted_set(R))
  return codim(P)
end


@attr Int function dim(R::MPolyLocRing{<:Field,<:Any,<:MPolyRing,<:MPolyRingElem, <:MPolyComplementOfKPointIdeal})
  # localization of a polynomial ring over a field at a maximal ideal does not change the dimension
  # because all maximal ideals have the same dimension in this case. 
  return dim(base_ring(R))
end
```

simonbrandhorst committed:
Update src/Rings/mpolyquo-localizations.jl
